core/src/pixelcolor/transparent_color.rs

```diff
@@ -0,0 +1,140 @@
+///         let alpha = ((inner & 0xFF00) >> 8) as f32 / 255.;
+///         TransparentGray { alpha, gray: Gray8::new((inner & 0xFF) as u8)}
+///     }
+/// }
+///
+/// // From<TransparentGray> is needed for PixelColor
+/// impl From<TransparentGray> for RawU16 {
+///     fn from(value: TransparentGray) -> Self {
+///         let TransparentGray { alpha, gray } = value;
+///         RawU16::new(((alpha * 255.) as u16) << 8 + gray.luma())
+///     }
+/// }
+///
+/// // Implement required methods
+/// impl TransparentColor<Gray8> for TransparentGray {
+///     fn blend_over(&self, other: Gray8) -> Gray8 {
+///         let luma = (self.alpha * self.gray.luma() as f32) + (1.-self.alpha) * other.luma() as f32;
+///         Gray8::new(luma as u8)
+///     }
+/// }
+///
+/// // Make conversion available for `DrawTargetExt::color_converted`
+/// impl From<Gray8> for TransparentGray {
+///     fn from(value: Gray8) -> Self {
+///         TransparentGray { alpha: 1., gray: value }
+///     }
+/// }
+///
+/// // Mock transparent handling display
+/// pub struct Display {}
+/// impl DrawTarget for Display {
+///     type Color = TransparentGray;
+///     type Error = core::convert::Infallible;
+///
+///     fn draw_iter<I>(&mut self, pixels: I) -> Result<(), core::convert::Infallible>
+///     where
+///         I: IntoIterator<Item = Pixel<Self::Color>>,
+///     {
+///         for Pixel(point, color) in pixels.into_iter() {
+///             // mock existing pixel as middle gray
+///             let current_color = Gray8::new(127);
+///             // blend the color with existing color
+///             let final_color = color.blend_over(current_color);
+///             // ... draw final color on display
+///         }
+///         Ok(())
+///     }
+/// }
+/// impl OriginDimensions for Display {
+///     fn size(&self) -> Size {
+///         Size::new(300, 300)
+///     }
+/// }
+///
+/// // The image data.
+/// const DATA: &[u8] = &[
+///     0x80, 0xFF, 0xFF, 0x80, //
+///     0xFF, 0x00, 0x00, 0xFF, //
+///     0xFF, 0x00, 0x00, 0xFF, //
+///     0x80, 0xFF, 0xFF, 0x80, //
+/// ];
+///
+/// // Create a 4x4 non-transparent image
+/// let raw_image = ImageRaw::<Gray8>::new(DATA, Size::new(4,4)).unwrap();
+/// let image = Image::new(&raw_image, Point::zero());
+///
+/// // Create a transparent handling display
+/// let mut display = Display{};
+///
+/// // Draw a non-transparent object on a transparent handling display
+/// // Conversion is automatically handled by the `From` trait above and the `DrawTargetExt` trait
+/// image.draw(&mut display.color_converted()).unwrap();
+/// ```
+pub trait TransparentColor<C: PixelColor=Self>: PixelColor {
+    /// Blend this color over another underlying color
+    fn blend_over(&self, other: C) -> C;
+}
+
+/// Trait for transparent colors that expose their alpha channel
+pub trait AlphaColor: TransparentColor<Self::OpaqueColor> {
+    /// Alpha channel type
+    type Alpha;
```

Comment:
Why did you add this associated type? Is it to allow a `bool` alpha channel or a bit depth > 8?

I don't think that an > 8 bits alpha channel is necessary if `RgbColor` is limited to 8 bits per channel max. While a `bool` alpha channel could be nice for on-off transparency, it might make the implementation of generic conversions between color types more difficult. And the `MAX_A` constant doesn't really make sense for a bool.

Reply:
The `Alpha` associated type is because some people use float as alpha channel, but it might be useless in embedded-graphics.

The `MAX_A` is to mimic the behavior of `RgbColor` colors and allow colors like `Argb6666`, but see that it's not compatible with an associated Alpha type.

Reply:
> The `Alpha` associated type is because some people use float as alpha channel, but it might be useless in embedded-graphics.

Interesting, I hadn't thought about floating point values at all. Without also supporting floats for the RGB channels this might not be of much use, but perhaps a conversion function or special constructor could be a useful addition later. Specifying alpha as a float from 0 to 1 is much easier that having to think about how many alpha channel bits the current color type has and scaling the value manually.

Reply:
It is easier ton understand, but if the color is based on integer, it needs a lot of conversions for blending.